File: leetcode/ShortestWaytoFormString/20201121.a.cpp
/* Copyright (C) 2019 by iamslash */

#include <cstdio>
#include <string>

// s: abc
//       i
// t: abcbc
//         j
// a: 1

// s: abc
//    i
// t: acdbc
//      j
// a: 3

// O(T) O(1)
class Solution {
 public:
  int shortestWay(string s, string t) {
    int i = 0, j = 0, ans = 0;
    while (j < t.size() && ans <= j) {
      while (s[i] != t[j] && i < s.size())
        ++i;
      if (i == s.size()) {
        i = 0;
        ++ans;
      } else {
        ++i;
        ++j;
      }
    }
    return ans <= j ? ans + 1 : -1;
  }
};

int main() {
  return 0;
}

File: leetcode/ShortestWaytoFormString/20201121.a.js
// Copyright (C) 2020 by iamslash

//   s: a b c
//            i
//   t: a b c b c
//                j
// ans: 1

// 84ms 31.28% 39MB 9.48%
// greedy algorithm
// O(T) O(1)
var shortestWay = function(src, dst) {
    let i = 0, j = 0, ans = 0, m = src.length, n = dst.length;
    while (j < n && ans <= j) {
        while (src[i] != dst[j] && i < m) {
            ++i;
        }
        if (i == m) {
            i = 0;
            ++ans;
        } else {
            ++i;
            ++j;
        }
    }
    if (ans > j) {
        return -1
    }
    return ans + 1;
};
